aptos-move/aptos-vm-types/src/resource_group_adapter.rs

```diff
@@ -0,0 +1,525 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::resolver::{TResourceGroupView, TResourceView};
+use anyhow::Error;
+use aptos_state_view::TStateView;
+use aptos_types::{
+    access_path::AccessPath, on_chain_config::ConfigStorage, state_store::state_key::StateKey,
+};
+use bytes::Bytes;
+use move_core_types::{language_storage::StructTag, value::MoveTypeLayout};
+use serde::{de::DeserializeOwned, Serialize};
+use std::{
+    cell::RefCell,
+    collections::{BTreeMap, HashMap},
+    fmt::Debug,
+    hash::Hash,
+};
+
+/// Corresponding to different gas features, methods for counting the 'size' of a
+/// resource group. None leads to 0, while AsBlob provides the group size as the
+/// size of the serialized blob of the BTreeMap corresponding to the group.
+/// For AsSum, the size is summed for each resource contained in the group (of
+/// the resource blob, and its corresponding tag, when serialized)
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum GroupSizeKind {
+    None,
+    AsBlob,
+    AsSum,
+}
+
+impl GroupSizeKind {
+    pub fn from_gas_feature_version(gas_feature_version: u64) -> Self {
+        if gas_feature_version >= 9 {
+            if gas_feature_version >= 12 {
+                GroupSizeKind::AsSum
+            } else {
+                // Keep old caching behavior for replay.
+                GroupSizeKind::AsBlob
+            }
+        } else {
+            GroupSizeKind::None
```

Comment:
In what scenario do way not account for the group size? I would assume in that case the blocb size is still charged for bytes? (Just to make sure)

Reply:
The charge is the bytes size, plus for the first access from some gas feature version it adds the resource group size (as a blob). With next version it changes again, and the 3 variants correspond to these gas feature behaviors.

Reply:
So we used to charge only the individual resource size but not the group? I hope we don't charge `group_size + individual_resource_size` in any case 😂

Reply:
LoL, you should know?
I think it was just resource size, nothing about group for a while. Then, it changes to add group (serialized blob) size on the first access, so on that access it's like that sum. And then (after first group access), it is just the resource size.

Reply:
Recent PRs change the group size computation, in a minor way (instead of the whole as blob, sum over elements), but nothing else.